src/cbs.c

```diff
@@ -167,57 +356,102 @@ int cbs_close(CBS_HANDLE cbs)
 
     if (cbs == NULL)
     {
+        /* Codes_SRS_CBS_01_045: [ If the argument `cbs` is NULL, `cbs_close` shall fail and return a non-zero value. ]*/
+        LogError("NULL cbs handle");
+        result = __FAILURE__;
+    }
+    else if (cbs->cbs_state == CBS_STATE_CLOSED)
+    {
+        /* Codes_SRS_CBS_01_047: [ `cbs_close` when closed shall fail and return a non-zero value. ]*/
+        /* Codes_SRS_CBS_01_048: [ `cbs_close` when not opened shall fail and return a non-zero value. ]*/
+        LogError("Already closed");
         result = __FAILURE__;
     }
     else
     {
+        if (cbs->cbs_state == CBS_STATE_OPENING)
+        {
+            /* Codes_SRS_CBS_01_079: [ If `cbs_close` is called while OPENING, the `on_cbs_open_complete` callback shall be called with `CBS_OPEN_CANCELLED`, while passing the `on_cbs_open_complete_context` as argument. ]*/
+            cbs->on_cbs_open_complete(cbs->on_cbs_open_complete_context, CBS_OPEN_CANCELLED);
+        }
+
+        /* Codes_SRS_CBS_01_044: [ `cbs_close` shall close the CBS instance by calling `amqpmanagement_close` on the underlying AMQP management handle. ]*/
         if (amqpmanagement_close(cbs->amqp_management) != 0)
         {
+            /* Codes_SRS_CBS_01_046: [ If `amqpmanagement_close` fails, `cbs_close` shall fail and return a non-zero value. ]*/
+            LogError("Failed closing AMQP management instance");
             result = __FAILURE__;
         }
         else
         {
+            cbs->cbs_state = CBS_STATE_CLOSED;
+
+            /* Codes_SRS_CBS_01_080: [ On success, `cbs_close` shall return 0. ]*/
             result = 0;
         }
     }
 
     return result;
 }
 
-int cbs_put_token(CBS_HANDLE cbs, const char* type, const char* audience, const char* token, ON_CBS_OPERATION_COMPLETE on_operation_complete, void* context)
+int cbs_put_token_async(CBS_HANDLE cbs, const char* type, const char* audience, const char* token, ON_CBS_OPERATION_COMPLETE on_cbs_put_token_complete, void* on_cbs_put_token_complete_context)
 {
     int result;
 
+    /* Codes_SRS_CBS_01_083: [ `on_cbs_put_token_complete_context` shall be allowed to be NULL. ]*/
     if ((cbs == NULL) ||
-        (token == NULL))
+        (type == NULL) ||
+        (audience == NULL) ||
+        (token == NULL) ||
+        (on_cbs_put_token_complete == NULL))
     {
+        /* Codes_SRS_CBS_01_050: [ If any of the arguments `cbs`, `type`, `audience`, `token` or `on_cbs_put_token_complete` is NULL `cbs_put_token_async` shall fail and return a non-zero value. ]*/
+        LogError("Bad arguments: cbs = %p, type = %p, audience = %p, token = %p, on_cbs_put_token_complete = %p",
+            cbs, type, audience, token, on_cbs_put_token_complete);
+        result = __FAILURE__;
+    }
+    else if ((cbs->cbs_state == CBS_STATE_CLOSED) ||
+        (cbs->cbs_state == CBS_STATE_ERROR))
```

Comment:
You are so freaking smart!
Yes, I wanted to actually block put_token and delete_token in OPENING, but that would actually break IoTHub client and I would need lots of changes there to deal with it. WHich I did not quite want at this point...

---
In reply to: [107039495](https://github.com/Azure/azure-uamqp-c/pull/56#discussion_r107039495) [](ancestors = 107039495)